### PR TITLE
feat: add maxRetry to taskSpec

### PR DESCRIPTION
**Pull request by merryzhou**
Signed-off-by: merryzhou <merryzhou0882@gmail.com>

https://github.com/volcano-sh/volcano/issues/1099

### CODE REVIEW

**Comment by Thor-wl** [comment]
@merryzhou Please resolve the confliction

**Comment by merryzhou** [comment]
> @merryzhou Please resolve the confliction

done

**Comment by Thor-wl** [comment]
Because of the separation for the pkg/apis, this PR should be updated. @merryzhou

**Comment by LuBingtan** [comment]
@merryzhou 
Hi，any update about this PR?
I found this PR https://github.com/volcano-sh/apis/pull/42 for volcano-sh/apis
Are you still working on this ?

**Comment by merryzhou** [comment]
@Thor-wl @william-wang PTAL

File: pkg/controllers/cache/cache.go [hunk 1]
```diff
@@ -284,6 +284,58 @@ func (jc *jobCache) TaskCompleted(jobKey, taskName string) bool {
...
+
+	taskPods, found := jobInfo.Pods[taskName]
+
+	if !found {
+		return false
+	}
+
+	if jobInfo.Job == nil {
+		return false
+	}
```
Comment (Thor-wl) [inline pkg/controllers/cache/cache.go:306]:
```
   if !found || jobInfo.Job == nil {
       return false
   }
```
Conditions merged may be better?

File: pkg/controllers/cache/cache.go [hunk 2]
```diff
@@ -284,6 +284,58 @@ func (jc *jobCache) TaskCompleted(jobKey, taskName string) bool {
...
+
+	for _, task := range jobInfo.Job.Spec.Tasks {
+		if task.Name == taskName {
+			maxRetry = task.MaxRetry
+			taskReplicas = task.Replicas
+			break
+		}
+	}
+	if taskReplicas <= 0 || maxRetry == -1 {
```
Comment (Thor-wl) [inline pkg/controllers/cache/cache.go:315]: Why `taskReplicas ` can be less than 0 and `maxRetry` will be -1?

File: pkg/controllers/cache/cache.go [hunk 3]
```diff
@@ -284,6 +284,58 @@ func (jc *jobCache) TaskCompleted(jobKey, taskName string) bool {
...
+				stat := pod.Status.InitContainerStatuses[j]
+				retried += stat.RestartCount
+			}
+			for j := range pod.Status.ContainerStatuses {
+				stat := pod.Status.ContainerStatuses[j]
+				retried += stat.RestartCount
+			}
+		}
+	}
+	return retried >= maxRetry
```
Comment (Thor-wl) [inline pkg/controllers/cache/cache.go:336]: `retried == maxRetry` should be considered as `TaskFailed`?

File: test/e2e/util/job.go [hunk 4]
```diff
@@ -211,10 +212,16 @@ func CreateJobInner(ctx *TestContext, jobSpec *JobSpec) (*batchv1alpha1.Job, err
...
+		maxRetry := task.MaxRetry
+		if maxRetry == 0 {
+			maxRetry = -1
+		}
+
 		ts := batchv1alpha1.TaskSpec{
 			Name:     name,
 			Replicas: task.Rep,
 			Policies: task.Policies,
+			MaxRetry: maxRetry,
```
Comment (Thor-wl) [inline test/e2e/util/job.go:224]: Suggest to supply some introduction and examples about this field to make users and developers have a better understanding about that.

File: pkg/controllers/cache/cache.go (hunk 2 above)
Comment (merryzhou) [reply pkg/controllers/cache/cache.go]: maxRetry == -1 means no retry limit；

File: pkg/controllers/cache/cache.go (hunk 1 above)
Comment (merryzhou) [reply pkg/controllers/cache/cache.go:302]: done

File: test/e2e/util/job.go (hunk 4 above)
Comment (merryzhou) [reply test/e2e/util/job.go:224]: already added e2e: "job level LifecyclePolicy, Event: TaskFailed; Action: TerminateJob"

File: pkg/controllers/cache/cache.go (hunk 2 above)
Comment (Thor-wl) [reply pkg/controllers/cache/cache.go]: > maxRetry == -1 means no retry limit；

OK. If that it's better to add some comments

**Review by jasonliu747 — COMMENTED** [review]
/lgtm overall, others PTAL.
Also, please fix the CI error, thanks.

File: test/e2e/util/job.go (hunk 4 above)
Comment (Thor-wl) [reply test/e2e/util/job.go:224]: > already added e2e: "job level LifecyclePolicy, Event: TaskFailed; Action: TerminateJob"

THX. Perhaps add a documentation under  docs/ is better.

**Review by Thor-wl — APPROVED** [review]

**Pull request merged**